app/Tasks/Types/Types.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+namespace App\Tasks\Types;
+
+use App\Models\Task;
+
+/**
+ * Abstract class for task types
+ * NOTE: Experimental feature, this interface/class is not stable. Changes to architecture may occur.
+ */
+abstract class Types
+{
+    protected $name;
+
+    protected $icon;
+
+    public function __construct()
+    {
+        $this->name = $this->getName();
+        $this->icon = $this->getIcon();
+    }
+
+    public function onCreated(Task $model)
+    {
+        // Default behaviour when task is created
+    }
+
+    public function onCompleted(Task $model)
+    {
+        // Default behaviour is completed
+    }
+
+    public function onDeclined(Task $model)
+    {
+        // Default behaviour is declined
+    }
```

Comment:
thought (out of scope, re: abstarct vs implementation): here's where I suspect we could define each of these behaviours as interfaces, which task types could implement, letting our dispatcher check if the task supports a given hook when a task is e.g. completed, before then calling the task. That'd give a degree of explicitness between the controller and the specific task types, and you wouldn't rely on inherited behaviour.

Reply:
This is indeed something we should look at in the next iteration. One thought I can add is: Implements classes iirc require you to override all of the defined functions, meaning we can't have a default implementation of what happens when any type of issue is closed. However, could be we don't really need it either, hence it's empty today in your highlighted comment. 

The future use and ideas will shine a light of the best implementation of the architecture for the next iteration at least ⭐